text/3503-frontmatter.md

```diff
@@ -0,0 +1,546 @@
+- Unfamiliar, new syntax, unclear how it will work out for newer users
+
+### Alternative 5: Regular Comment
+
+Simple header:
+```rust
+#!/usr/bin/env cargo
+/* Cargo.toml:
+[package]
+edition = "2018"
+*/
+
+fn main() {
+}
+```
+
+HEREDOC:
+```rust
+#!/usr/bin/env cargo
+/* Cargo.toml >>>
+[package]
+edition = "2018"
+<<<
+*/
+
+fn main() {
+}
+```
+The manifest can be a regular comment with a header.  If we were to support
+multiple types of content (manifest, lockfile), we could either use multiple
+comments or HEREDOC.
+This does not prescribe the exact syntax used or supported comments
+
+Benefits
+- Natural to support `Cargo.lock` as well
+- Without existing baggage, can use file extensions, making a firmer
+  association in users minds for what is in these (for those used to `Cargo.toml`)
+- Depending on the exact syntax decided on, users may be able to edit/copy/paste the manifest without dealing with leading characters
+
+Downsides
+- Unfamiliar syntax
+- When discussing with a Rust crash course teacher, it was felt their students would have a hard time learning to write these manifests from scratch
+  - Having the explain the overloading of concepts to new users
+  - Unpredictable location (both the doc comment and the cargo code block within it)
+  - Visual clutter (where clutter is overwhelming already in Rust)
+- New style of structured comment for the ecosystem to support with potential
+  compatibility issues, likely requiring a new edition
+- Assuming it can't be parsed with `syn` and either we need to write a
+  sufficiently compatible comment parser or pull in a much larger rust parser
+  to extract and update comments.
+  - Like with doc comments, this should map to an attribute and then we'd just start the MVP with that attribute
+
+### Alternative 6: Static-site generator frontmatter
+
+```rust
+#!/usr/bin/env cargo
+---
+[package]
+edition = "2018"
+---
+
+fn main() {
+}
+```
+This is a subset/specialization of YAML presentation streams that mirrors people's experience with static site generators:
+- The first line post-shebang-stripping is 3+ dashes, then capture all content until a matching pair of dashes on a dedicated line.  This would be captured into a `#![frontmatter = ""]`.  `frontmatter` attribute is reserved for crate roots.  The 3+ with matching pair is a "just in case" a TOML multi-line string has that syntax in it)
+- Future evolution: Allow a markdown-like infostring on the frontmatter opening dashes to declare the format with `cargo` being the default
+- Future evolution: Allow `frontmatter` attribute on any module
+
+Benefits
+- Visually/syntactically lightweight
+- Has parallels to ideas outside of Rust, building on external knowledge that might exist
+- Easy for cargo to parse and modify
+- Can easily be leveraged by buck2, meson, etc in the future
+- Users can edit/copy/paste the manifest without dealing with leading characters
+
+Downsides
+- Too general that people might abuse it
```

Comment:
As a general comment.. I don't agree with this as a downside. I'm not even sure what you mean by "abuse" since that varies greatly by use case. Will people embed 1,200 source files in a single file? In general probably not, but if they do, they probably have a good reason.

For example, if I want to run a minimizer like creduce (which supports Rust) to reproduce a compiler issue, it helps to embed my reproducer in a single file. Some tooling assumes this because C/C++ compilation units can always be embedded in a single file, unlike Rust. Then the tool can take care of minimizing for me. Obviously I wouldn't do this for every day software development, but in my mind it's a completely valid use case.